### PR TITLE
Refactor the Receive Transaction Protocol into the Protocol pattern

### PR DESCRIPTION
**Pull request by philipr-za**
## Description
This PR refactors the Receive Transaction Protocol in the Transaction Service. It moves all the logic for handling accepting a transaction, sending the reply and waiting for the finalised transaction into a self contained Protocol pattern. This protocol was added to the Transaction Service and plumbed in as with the previous protocols.

All the send operations in the Send and Receive transaction protocols were updated to follow the strategy of attempt direct and only send via SAF if discovery is needed or the direct send times out.

It was also decided that the Send and Receive transaction protocols should not monitor the liveness of the counterparty for resending purposes so that was removed from the Send Transaction Protocol. This PR also removes the `test_resend_of_tx_on_pong_event` which tested this functionality.

The Send Transaction Protocol was also cleaned up to better reflect the stages it can be started in.

A test was added to test that both the send and receive transaction protocols restart when the wallet starts up again.

## How Has This Been Tested?
Tests updated and new test provided that tests that the protocols restart on startup.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs [hunk 1]
```diff
@@ -101,80 +102,95 @@ where TBackend: TransactionBackend + Clone + 'static
...
-            let tx_id = msg.tx_id;
+        match self.stage {
+            TransactionSendProtocolStage::Initial => {
+                self.initial_send_transaction().await?;
+                self.wait_for_reply().await?;
+            },
+            TransactionSendProtocolStage::WaitForReply => {
+                self.wait_for_reply().await?;
+            },
+            TransactionSendProtocolStage::WaitForFinalizeAck => panic!("FARK"),
```
Comment (sdbondi) [inline base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs]: 🤣

**Review by sdbondi — DISMISSED** [review]
LGTM - just the 🖇️

File: base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs (hunk 1 above)
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs]: lol, woops that should have come out

**Comment by philipr-za** [comment]
Sorted out the clippy stuff.

**Review by sdbondi — APPROVED** [review]

**Pull request merged**